platform/wayland/cog-platform-wl.c

```diff
@@ -281,22 +339,248 @@ static struct {
+
+    xdg_surface_add_listener(win_data.shell_context->xdg_shell_data.xdg_surface, &xdg_surface_listener, NULL);
+    win_data.shell_context->xdg_shell_data.xdg_toplevel =
+        xdg_surface_get_toplevel(win_data.shell_context->xdg_shell_data.xdg_surface);
+    g_assert(win_data.shell_context->xdg_shell_data.xdg_toplevel);
+
+    xdg_toplevel_add_listener(win_data.shell_context->xdg_shell_data.xdg_toplevel, &xdg_toplevel_listener, NULL);
+    xdg_toplevel_set_title(win_data.shell_context->xdg_shell_data.xdg_toplevel, COG_DEFAULT_APPNAME);
+
+    const char *  app_id = NULL;
+    GApplication *app = g_application_get_default();
+    if (app) {
+        app_id = g_application_get_application_id(app);
+    }
+    if (!app_id) {
+        app_id = COG_DEFAULT_APPID;
+    }
+    xdg_toplevel_set_app_id(win_data.shell_context->xdg_shell_data.xdg_toplevel, app_id);
+    wl_surface_commit(win_data.shell_context->wl_shell_data.wl_surface);
+}
+static void
+create_xdg_shell_popup()
+{
+    popup_data.xdg_positioner = xdg_wm_base_create_positioner(wl_data.xdg_shell);
+    g_assert(popup_data.xdg_positioner);
+
+    xdg_positioner_set_size(popup_data.xdg_positioner, popup_data.width, popup_data.height);
+    xdg_positioner_set_anchor_rect(popup_data.xdg_positioner, 0, (win_data.height - popup_data.height),
+                                   popup_data.width, popup_data.height);
+
+    popup_data.shell_context->xdg_shell_data.xdg_surface =
+        xdg_wm_base_get_xdg_surface(wl_data.xdg_shell, popup_data.shell_context->wl_shell_data.wl_surface);
+    g_assert(popup_data.shell_context->xdg_shell_data.xdg_surface);
+
+    xdg_surface_add_listener(popup_data.shell_context->xdg_shell_data.xdg_surface, &xdg_surface_listener, NULL);
+    popup_data.shell_context->xdg_shell_data.xdg_popup =
+        xdg_surface_get_popup(popup_data.shell_context->xdg_shell_data.xdg_surface,
+                              win_data.shell_context->xdg_shell_data.xdg_surface, popup_data.xdg_positioner);
+    g_assert(popup_data.shell_context->xdg_shell_data.xdg_popup);
+
+    xdg_popup_add_listener(popup_data.shell_context->xdg_shell_data.xdg_popup, &xdg_popup_listener, NULL);
+    xdg_popup_grab(popup_data.shell_context->xdg_shell_data.xdg_popup, wl_data.seat, wl_data.event_serial);
+    wl_surface_commit(popup_data.shell_context->wl_shell_data.wl_surface);
+}
+
+static shell_functions xdg_shell_functions = {.enter_fullscreen = &xdg_shell_enter_fullscreen,
+                                              .exit_fullscreen = &xdg_shell_exit_fullscreen,
+                                              .destroy_shell = &destroy_xdg_shell,
+                                              .maximize_surface = &maximize_xdg_shell_surface,
+                                              .create_window = &create_xdg_shell_window,
+                                              .create_popup = &create_xdg_shell_popup};
+xdg_shell_data         xdg_data = {.functions = &xdg_shell_functions};
```

Comment:
If functions wasn't a pointer, you could just declare the functions struct directly here;
```
.functions = { .enter_fullscreen = ...,
                      ... }
```